src/java.base/share/classes/java/util/LazyCollections.java

```diff
@@ -0,0 +1,650 @@
+            final Object candidate = new Object();
+            final Object witness = UNSAFE.compareAndExchangeReference(mutexes, offset, null, candidate);
+            return witness == null ? candidate : witness;
+        }
+
+        private void releaseMutex(long offset) {
+            UNSAFE.putReference(mutexes, offset, TOMB_STONE);
+            if (counter != null && counter.decrementAndGet() == 0) {
+                mutexes = null;
+                counter = null;
+            }
+        }
+
+    }
+
+    @ForceInline
+    private static long offsetFor(long index) {
+        return Unsafe.ARRAY_OBJECT_BASE_OFFSET + Unsafe.ARRAY_OBJECT_INDEX_SCALE * index;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <E> E[] newGenericArray(int length) {
+        return (E[]) new Object[length];
+    }
+
+    public static <E> List<E> ofLazyList(int size,
+                                         IntFunction<? extends E> computingFunction) {
+        return new LazyList<>(size, computingFunction);
+    }
+
+    public static <K, V> Map<K, V> ofLazyMap(Set<K> keys,
+                                             Function<? super K, ? extends V> computingFunction) {
+        return new LazyMap<>(keys, computingFunction);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <K, E extends Enum<E>, V>
+    Map<K, V> ofLazyMapWithEnumKeys(Set<K> keys,
+                                    Function<? super K, ? extends V> computingFunction) {
+        // The input set is not empty
+        final Class<E> enumType = ((E) keys.iterator().next()).getDeclaringClass();
+        final BitSet bitSet = new BitSet(enumType.getEnumConstants().length);
+        int min = Integer.MAX_VALUE;
+        int max = Integer.MIN_VALUE;
+        for (K t : keys) {
+            final int ordinal = ((E) t).ordinal();
+            min = Math.min(min, ordinal);
+            max = Math.max(max, ordinal);
+            bitSet.set(ordinal);
+        }
+        final int backingSize = max - min + 1;
+        final IntPredicate member = ImmutableBitSetPredicate.of(bitSet);
+        return (Map<K, V>) new LazyEnumMap<>((Set<E>) keys, enumType, min, backingSize, member, (Function<E, V>) computingFunction);
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> T orElseComputeSlowPath(final T[] array,
+                                       final int index,
+                                       final Mutexes mutexes,
+                                       final Object input,
+                                       final FunctionHolder<?> functionHolder) {
+        final long offset = offsetFor(index);
+        final Object mutex = mutexes.acquireMutex(offset);
+        preventReentry(mutex);
+        synchronized (mutex) {
+            final T t = array[index];  // Plain semantics suffice here
+            if (t == null) {
+                final T newValue = switch (functionHolder.function()) {
+                    case Supplier<?> sup     -> (T) sup.get();
```

Comment:
Is the held function ever a Supplier? I don't see a FunctionHolder being created with one.

Reply:
This is from when `FunctionHolder` was located in the `jdk.internal.lang.stable` package in [GH‑25878], and was used to implement the computing function freeing mechanism in [`StableValue​::supplier(…)`].

[GH‑25878]: https://github.com/openjdk/jdk/pull/25878
[`StableValue​::supplier(…)`]: https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/lang/StableValue.html#supplier(java.util.function.Supplier)